### PR TITLE
[#37]  added BailSyntaxErrorStrategy

### PR DESCRIPTION
**Pull request by xnqio**
* added a new ErrorStrategy to print out GDL syntax errors while still throwing an exception in case of the input not matching the expected form
* this newly created `BailSyntaxErrorStrategy` behaves like a mixture of the `DefaultErrorStrategy` and the `BailErrorStrategy`

### CODE REVIEW

**Comment by galpha** [comment]
Hi @s1ck. If you give your approval to this PR, I can proceed and prepare the next release.

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import org.antlr.v4.runtime.DefaultErrorStrategy;
+import org.antlr.v4.runtime.InputMismatchException;
+import org.antlr.v4.runtime.Parser;
+import org.antlr.v4.runtime.ParserRuleContext;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Token;
+import org.antlr.v4.runtime.misc.ParseCancellationException;
+
+/**
+ * This ErrorStrategy throws an exception if the syntax is not matching the expected GDL language, but, in
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
 * This ErrorStrategy throws an exception if the syntax is not matching the expected GDL language, but in
```

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+   *  {@link DefaultErrorStrategy#recover(Parser, RecognitionException)} method
+   *  get executed
+   */
+  @Override
+  public void recover(Parser recognizer, RecognitionException e) {
+    super.recover(recognizer, e);
+    for (ParserRuleContext context = recognizer.getContext(); context != null; context = context.getParent()) {
+      context.exception = e;
+    }
+
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
```

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+ * addition to the BailErrorStrategy also prints out the first syntax mismatch
+ */
+public class BailSyntaxErrorStrategy extends DefaultErrorStrategy {
+  /**
+   *  Instead of recovering from exception {@code e}, re-throw it wrapped
+   *  in a {@link ParseCancellationException} so it is not caught by the
+   *  rule function catches.  Use {@link Exception#getCause()} to get the
+   *  original {@link RecognitionException}. To print the syntax error the
+   *  {@link DefaultErrorStrategy#recover(Parser, RecognitionException)} method
+   *  get executed
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
   *  gests executed.
```

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  /**
+   *  Make sure we don't attempt to recover inline; if the parser
+   *  successfully recovers, it won't throw an exception.
+   *  Again, the {@link DefaultErrorStrategy#recoverInline(Parser)} gets executed
+   *  to print the wrong syntax
+   */
+  @Override
+  public Token recoverInline(Parser recognizer)
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
  public Token recoverInline(Parser recognizer) throws RecognitionException {
```

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+    throws RecognitionException
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
```

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  {
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
```

File: src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    super.recoverInline(recognizer);
+    InputMismatchException e = new InputMismatchException(recognizer);
+    for (ParserRuleContext context = recognizer.getContext(); context != null; context = context.getParent()) {
+      context.exception = e;
+    }
+
```
Comment (s1ck) [inline src/main/java/org/s1ck/gdl/exceptions/BailSyntaxErrorStrategy.java]:
```suggestion
```

**Comment by galpha** [comment]
Since I still have no write access to the feature branch I'll add your suggestions in the following pr.

**Pull request merged**